minerl/env/observations.py

```diff
@@ -0,0 +1,116 @@
+            os.path.isfile(os.path.join(dir_path, 'data', 'items.txt')) and
+            os.path.isfile(os.path.join(dir_path, 'data', 'blocks.txt'))):
+        print("Could not find item and block data!")
+        return None
+    item = []
+    block = []
+    with open(os.path.join(dir_path, 'data', 'items.txt')) as f:
+        for line in f:
+            item.append(line.strip())
+
+    with open(os.path.join(dir_path, 'data', 'blocks.txt')) as f:
+        for line in f:
+            block.append(line.strip())
+    return item, block
+
+
+ITEMS, BLOCKS = get_item_block_lists()
+# Use these for observation spaces that interact with minecraft items and blocks.
+#   Provides a non-meaningful but consistent ordering over the space of names.
+# Currently no support for entities.
+
+
+def pov_observation(info, obs_space):
+    """
+    :param info: Info dict generated by Malmo mod. Must contain 'pov'.
+    :param obs_space: unused.
+    :return: np.array with shape specified in Mission xml file.
+    """
+    return info['pov']
+
+
+def inventory_observation(info, obs_space):
+    """
+    For every item type in the observation space, add up how many of said item are in the inventory.
+    Duplicates the MineRL default inventory observation.
+
+    :param info: Info dict generated by Malmo mod.
+    :param obs_space: The observation space. Used to decide which item information to include.
+    :return: Dict of {item name : count in inventory}.
+    """
+    inventory_spaces = obs_space['inventory'].spaces
+
+    inventory_dict = {k: 0 for k in inventory_spaces}
+    # TODO change to maalmo
+    if 'inventory' in info:
+        for stack in info['inventory']:
+            if 'type' in stack and 'quantity' in stack:
+                type_name = stack['type']
+                if type_name == 'log2':
+                    type_name = 'log'
+
+                try:
+                    inventory_dict[type_name] += stack['quantity']
+                except ValueError:
+                    continue
+                except KeyError:
+                    # We only care to observe what was specified in the space.
+                    continue
+    else:
+        print("Inventory information could not be found, returning empty inventory.")
+
+    return inventory_dict
+
+
+def observation_creator(observation_space, info, extra_handlers=None):
+    """
+    Parses through the requested observation spaces and attempts to create each requested observation
+        from handlers defined in this file, using the given info dict.
+    
+    :param observation_space: Gym/MineRL observation space.
+    :param info: Python dict.
+    :param extra_handlers: Any extra functions to be used in decoding the observation space.
+    :return: Python dict with one index corresponding to each index in the observation space that has
+        an associated handler.
+    """
+    obs_space = deepcopy(observation_space.spaces)
+    obs_dict = {}
+
+    handlers = {
```

Comment:
The appeal here was having `handlers.update(...)` nicely take care of both including default handlers while letting them be overridden. I suppose dictionaries being mutable wouldn't be an issue here since presumably the env is the same across all calls, so the update method would be repeatedly called with the same extra handlers. Will do.